### ISSUE TITLE
Pull up and export lambdas in Annotations so that AoT Compiler can reference it.

### ISSUE THREAD

**Issue by mhevery** [angular/angular#13138]
Depends on: 
- [ ] Rewrite Tsickle to use TypeScript v2.1 transforms API see angular/tsickle#285

In JIT mode this code works.
```
@NgModule({
  providers: [
    {provide: SomeSymbol, useFactor: (i) => i.get('someSymbol'), deps: ['$injector']}
  ]
})
export class MyModule {}
```

but in AoT it needs to be written like so:
```
export factoryForSomeSymbol = (i) => i.get('someSymbol');

@NgModule({
  providers: [
    {provide: SomeSymbol, useFactor: factoryForSomeSymbol, deps: ['$injector']}
  ]
})
export class MyModule {}
```

`ngc` should be smart enough together with `tsickle` to perform this operation automatically for the developer.

**Comment by shlomiassaf** [comment]
@mhevery @tbosch @chuckjaz 

Now that Microsoft/TypeScript#13604 is in master, and will be in 2.2 this is really important for implementing the mixin pattern.

A use case will be an ActiveRecord library like ngResource but with full type information.

**Comment by chuckjaz** [comment]
@shlomiassaf I am uncertain how this issue is related to mixins and https://github.com/Microsoft/TypeScript/pull/13604. Can you go into more detail?

**Comment by shlomiassaf** [comment]
I misinterpreted the proposal, it wont help :/

Mixins are functions which are not supported by the compiler, not tsickle, mu bad.

**Comment by JoostK** [comment]
This has been implemented in b6c4af649505a7d158c7d4c0df4d43270bd29fd7 for Angular 5.
